### PR TITLE
Enhance diagonal (fixes #6479)

### LINKED ISSUE #6479: [feature request] enhance diagonal functionality
# Objective
PyTorch currently does not have a way to take diagonals with respect to arbitrary axes as e.g. `numpy.diagonal` has.
Two immediate use cases are
- Taking the diagonal of a batch of matrices. This is handy e.g. in Gaussian processes, but I have seen this a number of times.
- One of the missing `einsum` features (`torch.einsum('jii->ji', a)`) would benefit from this as well.
# Plan
I propose to do the following (see https://github.com/t-vi/pytorch/tree/diagonal_with_dim for a preliminary implementation)
- Implement `diagonal` natively in ATen instead of referring to TH/THC `diag`.
- Make `diagonal` return a view by adjusting shape, stride, and offset. `diag` copies, but this seems impractical for higher dimensional arrays and also unneeded. *This is not backward compatible.*
- Add two dimension parameters dim1 and dim2.
- Make a diagonal_backward method.
- Include the Tensor method, too.

*Edit*: Based on your feedback, I updated the code to use numpy semantics.

# ~~Limiting numpy compatibility?~~
~~NumPy's implementation has two features that might be worth to differ from:~~
- ~~NumPy's default axes are 0 and 1. I think it is more natural to use -2 and -1 ("batch thinking").~~
- ~~NumPy moves the new axis replacing the two old ones at the end of the tensor. I think it might be more natural to put it into the place of the first axis to be removed. I must admit that this is more a gut feeling than based on hard facts.~~

I would greatly appreciate your input on this, in particular regarding the potential numpy deviations.

### PR DESCRIPTION
**Pull request by t-vi**
This patch
- adds Tensor.diagonal to complement torch.diagonal
- implements diagonal natively in ATen
- makes diagonal a view (similar to numpy semantics, numpy currently has a read-only view to ease transition from diagonal returning a copy)
- implements taking arbitrary diagonals with numpy semantics
- implements diagonal backward instead of referring
  to the (more limited) diag

There is some discussion in #6479 .

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -44,11 +44,42 @@ Tensor diagflat(const Tensor& self, int64_t offset) {
...
+  // Note that we invert +/- in the second to absorb the negative
+  // sign in the offset.
+  if (offset >= 0) {
+    diag_size = std::min(self.size(dim1), self.size(dim2)-offset);
+    storage_offset += offset * self.stride(dim2);
+  } else {
+    diag_size = std::min(self.size(dim1)+offset, self.size(dim2));
+    storage_offset -= offset * self.stride(dim1);
+  }
+  AT_ASSERT(diag_size > 0, "invalid diagonal offset %zd", offset); // the diagonal offset was too large in magnitude
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp]: The comment suggests that this could happen, but is this actually impossible (and just a sanity check)?

Comment (t-vi) [reply aten/src/ATen/native/TensorShape.cpp]: It can happen if the user supplies a diagonal that lies entirely outside the matrix, e.g. `torch.diagonal(torch.randn(3,3),3)`.

File: test/test_torch.py [hunk 2]
```diff
@@ -1909,6 +1909,26 @@ def _test_diagonal(self, dtype, device):
     def test_diagonal(self):
         self._test_diagonal(self, dtype=torch.float32, device='cpu')
 
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_diagonal_multidim(self):
+        x = torch.randn(10, 11, 12, 13)
```
Comment (fmassa) [inline test/test_torch.py]: Can you also add a test where x is non-contiguous, like a permutation of what you have now?

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -717,6 +717,13 @@ Tensor diag_backward(const Tensor & grad, IntList input_sizes, int64_t diagonal)
   return grad_input;
 }
 
+Tensor diagonal_backward(const Tensor & grad, IntList input_sizes, int64_t offset, int64_t dim1, int64_t dim2) {
+  auto grad_input = at::zeros(grad.type(), input_sizes);
+  auto diag = at::diagonal(grad_input, offset, dim1, dim2);
+  diag.copy_(grad);
```
Comment (fmassa) [inline tools/autograd/templates/Functions.cpp]: Question: does this work with double backprop, as this is an in place ok?

Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: `gradgradcheck(lambda x: torch.diagonal(x,1,2), inputs=(torch.randn(2,3,4),))` returns True. Does that count as working?

Comment (fmassa) [reply tools/autograd/templates/Functions.cpp]: Should be fine then. I asked because I got bitten in the past with I place operations bring backwards, but I think that by default grad_input doesn't require grad so the double backprop should work fine

Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: No, it doesn't work.
```
a = torch.randn(5, 6, requires_grad=True)
b = torch.diagonal(a)**2
c = b.sum()
d, = torch.autograd.grad(c,a, retain_graph=True, create_graph=True)
```
now `d.requires_grad` will be zero.
The strange part is that if I exactly copypaste my diagonal code and replace `self` with `grad_in`, it suddenly works (I got the idea that it would work from the non-square case in `diag_backward`).
I have no idea what the difference is between `auto diag = grad_in.diagonal(...)` and doing the calculations locally and `auto diag = grad_in.as_strided(...)`.
How should I deal with that? Move the calculations in a utility function in a header?

Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: I'll duplicate the code for now and add a test that d.requires_grad = True in the above.

**Review by ezyang — APPROVED** [review]

File: test/test_torch.py [hunk 4]
```diff
@@ -1909,6 +1909,41 @@ def _test_diagonal(self, dtype, device):
...
+        expected = torch.diagonal(x, 2, 0, 1)
+        self.assertEqual(expected, result)
+        result = torch.diagonal(x, -2, 1, 2)
+        expected = xn.diagonal(-2, 1, 2)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
+        result = torch.diagonal(x, 0, -2, -1)
+        expected = xn.diagonal(0, -2, -1)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
```
Comment (apaszke) [inline test/test_torch.py]: Also, this basically looks like a few checks that do the same, except on different cases. Can you please replace those with a `for` loop over the possible cases? It will make code much much more readable.

File: test/test_torch.py [hunk 5]
```diff
@@ -1909,6 +1909,41 @@ def _test_diagonal(self, dtype, device):
...
+        # test non-continguous
+        xp = x.permute(1, 2, 3, 0)
+        result = torch.diagonal(xp, 0, -2, -1)
+        expected = xp.numpy().diagonal(0, -2, -1)
+        self.assertEqual(expected.shape, result.shape)
+        self.assertTrue(np.allclose(expected, result.numpy()))
+        # test that the backward requires grad
+        # we do this is because diagonal_backward uses inplace
+        # operations and gradgradcheck does not catch whether
+        # they works as expected
```
Comment (apaszke) [inline test/test_torch.py]: Why isn't `gradgradcheck` enough? I don't really understand the comment, but `test_torch` shouldn't check anything related to differentiation

File: tools/autograd/templates/Functions.cpp [hunk 6]
```diff
@@ -717,6 +717,40 @@ Tensor diag_backward(const Tensor & grad, IntList input_sizes, int64_t diagonal)
   return grad_input;
 }
 
+Tensor diagonal_backward(const Tensor & grad, IntList input_sizes, int64_t offset, int64_t dim1_, int64_t dim2_) {
+  auto grad_input = at::zeros(grad.type(), input_sizes);
+  // the following until the assignment of auto diag
+  // copies the diagonal code in aten/src/ATen/native/TensorShape.cpp
+  // that would be equivalent to
+  //        auto diag = grad_input.diagonal(offset, dim1, dim2);
+  // when using diagonal, the output is not differentiable twice
```
Comment (apaszke) [inline tools/autograd/templates/Functions.cpp]: Why wouldn't it be differentiable twice if you did that? It's generally ok to use an op to implement its own backward, and there are many cases where this happens: `transpose`, `permute`, `view`, ...

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -1305,6 +1313,17 @@
     -0.2239
     [torch.FloatTensor of size 2]
 
+    >>> x = torch.randn(2,5,4,2)
+    >>> torch.diagonal(x, -1, 1, 2)
```
Comment (apaszke) [inline torch/_torch_docs.py]: nit: Can you please add spaces between sizes in `randn`? Also, can you label the numbers with kwargs? The meaning of `-1, 1, 2` is not immediately obvious

File: tools/autograd/templates/Functions.cpp (hunk 6 above)
Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: The comment documents an empirical finding - I saw the implementation of diag_backward, thought "well, I can be smart about it and use diagonal rather than manually redoing the same thing" and it failed. I changed it to do the calculations here and it worked. I do not have a theory about why it happens. :neutral_face: I don't think it's using the op that is differentiated, but it is something subtle with the inplace operation.
This is also the reason why I think it is particularly important to check that it works in the tests.

File: test/test_torch.py (hunk 5 above)
Comment (t-vi) [reply test/test_torch.py]: Because gradgradcheck doesn't complain when the derivative does not require grad. So when I used the diagonal in the backward, it passed gradgradcheck, but the gradient was marked as not needing a derivative, and double backward was zero.
The test is a bit strange, can I move it to test_autograd.py then?

File: tools/autograd/templates/Functions.cpp (hunk 6 above)
Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: So here we have: https://github.com/pytorch/pytorch/pull/6718/files#diff-888d69f545029e79ccc606683a7ac08cR748
Working code:
```
  auto diag = grad_input.as_strided(sizes, strides, storage_offset);
  diag.copy_(grad);
  return grad_input;
```
Non-working code:
```
  auto diag = grad_input.diagonal(offset, dim1, dim2);
  diag.copy_(grad);
  return grad_input;
```
when `aten/src/ATen/native/TensorShape.cpp`'s `diagonal` returns a `self.as_strided(...)`.
So my best guess is that the first snipped has a shortcut in the automatically generated wrappers where the second id does not. I have never looked into how ATen tracks differentiability, so I would not know.

Mind you, I think I have reproduced this several times, but I might still have done something stupid, so if someone wants to check...

File: test/test_torch.py (hunk 5 above)
Comment (apaszke) [reply test/test_torch.py]: Well usually you just want this for all arguments, and it would improve the test coverage significantly, so I think it's worth it (not as part of this PR).

File: tools/autograd/templates/Functions.cpp (hunk 6 above)
Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: I found it! I need to register diagonal as a view function in gen_autograd.py at the top.
Thank you for insisting that I look into it - I have so much to learn.

File: test/test_torch.py (hunk 5 above)
Comment (t-vi) [reply test/test_torch.py]: OK, should I drop the check for now in anticipation of the increased test coverage or keep it with a comment?

**Comment by t-vi** [comment]
@ezyang is that my patch or the macos CI that has the bug? I didn't know that anything I did affected the fourier tests...

**Review by apaszke — APPROVED** [review]

**Comment by t-vi** [comment]
Hi, is there anything I can do to move this forward?
I'd be looking at adding the diagonal feature in einsum when the improved diagonal is available.

**Comment by ezyang** [comment]
@pytorchbot retest this please

I don't see how this PR could have triggered this. Let's try again.
